File: Two_Pointers/Python/125_valid_palindrome.py
# RUNTIME: 485 ms, faster than 5.17% of python3 Submissions. 
# MEMORY USAGE: 17.19 MB, less than 84.44% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Convert the string to lowercase
# 2. Remove all non-alphanumeric characters
# 3. Reverse the string
# 4. Compare the reversed string to the original string


class Solution:
    def isPalindrome(self, s: str) -> bool:
        def replaceAll(input):
            for c in input:
                if c in " ,.!?:;-_()[]{}\/|<>\#$%&=+*~^`@'\"":
                    input = input.replace(c, "")
            return input
        cropped_string = s.lower()
        cropped_string = replaceAll(cropped_string)
        reversed_string = cropped_string[::-1]
        return cropped_string == reversed_string

File: Two_Pointers/C++/125_valid_palindrome.cpp
/*
* RUNTIME: 6 ms - Beats 48.89% of submissions
* MEMORY - 9.31 MB - Beats 5.55% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
using namespace std;


/* Thoughtprocess:
* 1. Check if Char is alphanumeric
* 2. Convert Char to lowercase and store in seperate string
* 3. Reverse string
* 4. Compare strings
* 
* This could probaly be solved faster with starting from the middle and move left and right
*/


class Solution {
public:
    bool isPalindrome(string s) {

        string clearedStr = "";
        for (char c : s)
        {
            if ( (c >64 && c < 91) || (c > 96 && c < 123) || (c > 47 && c < 58))
            {
                clearedStr += tolower(c);
            }

        }
        string reversed = clearedStr;
        reverse(reversed.begin(), reversed.end());
        
        return clearedStr == reversed;

    }
};
